### PR TITLE
feat: Swap Fees used For Sybil Resistant Tx Fees  

### PR DESCRIPTION
**Pull request by lostak**
Closes: #1139
New PR: #1823 

## What is the purpose of the change

Swap fees paid for a msg are added to Tx fee when determining if the fee is sufficient for the gas price
- If the tx fee attached for a swap message is not sufficient to pay for gas, the swap fees paid can make up the difference
- The attached tx fee is always deducted in its entirety if it exists
- If the gas price is zero, no tx fee needs to be attached
- If the message satisfies the SybilResistantFee interface, add the swap fees paid to the sybil struct
- For all tx, add tx fee to sybil fees paid (for non-applicable msgs, feed paid = 0)
- Compare sybil fees paid to gas price when determining if a fee is sufficient

## Brief Changelog

- Add `GetSwapFee(ctx sdk.Context, poolId uint64) (sdk.Dec, error)` to `swap.go` and to the `GammKeeper` in the `expected_keepers` for the `txfees` module
- Add interface `SybilResistantFee` in a new file `msg_sybil.go` in `gamm/types/` 
- Add implement `SybilResistantFee` in all msgs involving a swap, including multihops
- Add file `sybil.go` for sybil resistant fees struct and methods
- `AnteHandle` now treats `SybilResistantFee` msgs differently when checking a fee but other msgs use the same logic as before

## Testing and Verifying

`make test`, `make test-sim` and `make build` are all successful but more tests are needed

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/gamm/types/msg_swap_sybil.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	SwapMsgRoute
+
+	PoolIdOnPath() []uint64
+	GetTokenToFee() sdk.Coin
+}
+
+var (
+	_ SybilResistantFeeSwap = MsgSwapExactAmountOut{}
+	_ SybilResistantFeeSwap = MsgSwapExactAmountIn{}
+)
```
Comment (lostak) [inline x/gamm/types/msg_swap_sybil.go:20]: Should this be expanded to include Join/Exit swap msgs?

Comment (lostak) [reply x/gamm/types/msg_swap_sybil.go]: Changes would be
https://github.com/osmosis-labs/osmosis/blob/6f00aa702e2e72ed20ab82cba87256b7a043307e/x/gamm/types/msg_swap_sybil.go#L10-L20
https://github.com/osmosis-labs/osmosis/blob/6f00aa702e2e72ed20ab82cba87256b7a043307e/x/gamm/types/msg_swap_sybil.go#L46-L58

File: x/gamm/types/msg_sybil.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+func (msg MsgJoinSwapShareAmountOut) PoolIdOnPath() (path []uint64) {
+	path[0] = msg.PoolId
+	return path
+}
+
+func (msg MsgJoinSwapShareAmountOut) TokenDenomsOnPath() (denom []string) {
+	denom[0] = msg.TokenInDenom
+	return denom
+}
+
```
Comment (lostak) [inline x/gamm/types/msg_sybil.go:77]: Should sysbil resistant fees be added for superfluid/incentivized pool delegate/undelegate?

**Comment by lostak** [comment]
[Relevant diagram depicting current implementation and PR's implementation](https://mm.tt/map/2327084301?t=14EgCd3vN7)

**Comment by mattverse** [comment]
Oh wow is the diagram created via software?

**Comment by lostak** [comment]
@mattverse no, did it own my own but I am confident in its accuracy

**Comment by lostak** [comment]
> Oh wow is the diagram created via software?

No, I created it on my own, but I am confident that it is accurate

**Review by alexanderbez — COMMENTED** [review]
There is a lot of commented out code here @lostak. What's up with that?

**Comment by lostak** [comment]
@alexanderbez I attempted to apply the same functionality for joins and exits involving swaps but ran into issues with the information provided in the messages. The commented code should be moved to another branch.

File: x/gamm/keeper/swap.go [hunk 3]
```diff
@@ -31,6 +31,15 @@ func (k Keeper) SwapExactAmountIn(
 	return k.swapExactAmountIn(ctx, sender, pool, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee)
 }
 
+func (k Keeper) GetSwapFeeFromPoolId(ctx sdk.Context, poolId uint64) (sdk.Dec, error) {
```
Comment (p0mvn) [inline x/gamm/keeper/swap.go:34]: Let's test this please

File: x/gamm/types/msg_sybil.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	_ SybilResistantFee = MsgSwapExactAmountIn{}
+)
+
+// MsgSwapExactAmountOut implements SybilResistantFee
+
+func (msg MsgSwapExactAmountOut) GetTokenToFee() sdk.Coin {
+	return msg.GetTokenOut()
+}
+
+func (msg MsgSwapExactAmountOut) PoolIdOnPath() []uint64 {
```
Comment (p0mvn) [inline x/gamm/types/msg_sybil.go:27]: I would like to see this covered by tests

File: x/txfees/keeper/sybil.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+func (s Sybil) AddToFeesPaid(feesPaid sdk.Coin) (Sybil, error) {
+	// Check same denom
+	if feesPaid.Denom != s.FeesPaid.Denom {
+		return Sybil{}, fmt.Errorf("Cannot add %s denom to sybil's %s fees paid denom", feesPaid.Denom, s.FeesPaid.Denom)
+	}
+	// Add tokens
+	fp := s.FeesPaid.Add(feesPaid)
+	// Return new sybil with tokens added together & previous gas price
+	return Sybil{GasPrice: s.GasPrice, FeesPaid: fp}, nil
```
Comment (p0mvn) [inline x/txfees/keeper/sybil.go:29]: What is the purpose for returning a new `Sybil`?

It seems to me that we can make `(s *Sybil)` (pointer receiver) and modify the existing instance without having to create new

File: x/txfees/keeper/feetokens.go [hunk 6]
```diff
@@ -50,6 +51,63 @@ func (k Keeper) CalcFeeSpotPrice(ctx sdk.Context, inputDenom string) (sdk.Dec, e
 	return spotPrice, nil
 }
 
+// GetTotalSwapFee gets the total swap fee for a swap message along a route of pool ids
```
Comment (p0mvn) [inline x/txfees/keeper/feetokens.go:54]:
```suggestion
// getTotalSwapFee gets the total swap fee for a swap message along a route of pool ids
```

File: x/txfees/keeper/feetokens.go [hunk 7]
```diff
@@ -50,6 +51,63 @@ func (k Keeper) CalcFeeSpotPrice(ctx sdk.Context, inputDenom string) (sdk.Dec, e
...
+	}
+	return swapFees, nil
+}
+
+func (k Keeper) feeTokenExists(ctx sdk.Context, denom string) bool {
+	return k.GetFeeTokensStore(ctx).Has([]byte(denom))
+}
+
+// getFeesPaid returns a token representing the fees paid along the route of swaps identified by the pool Ids
+func (k Keeper) getFeesPaid(ctx sdk.Context, poolIds []uint64, denomPath []string, token sdk.Coin) (sdk.Coin, error) {
```
Comment (p0mvn) [inline x/txfees/keeper/feetokens.go:85]: test please

**Review by p0mvn — CHANGES_REQUESTED** [review]
Hi @lostak 

Thanks so much for this change.

I haven't finished reviewing this in detail yet. I would like to request to test the newly added logic. Please let me know once that's ready

File: x/gamm/types/msg_sybil.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
+package types
```
[the remark this answers is not in the document]
Comment (lostak) [reply x/gamm/types/msg_sybil.go:1]: The reason I ended up moving them from `msg_swap.go` to `msg_sybil.go` was to keep `msg_swap.go` purely for swap messages, and to allow `msg_sybil.go` to be more abstract and be satisfied by msgs like join/exit with swap msgs.
If a join/exit with swap msg were to satisfy the methods of `SybilResistantFee`, then `TokenDenomsOnPath` would be implemented in `msg_sybil.go` but I understand your point. To that end, do you think `msg_swap.go` should define methods for msgs like `MsgJoinSwapExternAmountIn`?

File: x/txfees/keeper/feetokens.go [hunk 9]
```diff
@@ -50,6 +51,63 @@ func (k Keeper) CalcFeeSpotPrice(ctx sdk.Context, inputDenom string) (sdk.Dec, e
 	return spotPrice, nil
 }
 
+// GetTotalSwapFee gets the total swap fee for a swap message along a route of pool ids
+func (k Keeper) getTotalSwapFee(ctx sdk.Context, poolIds []uint64, denomPath []string) (sdk.Dec, error) {
+	prefixStore := k.GetFeeTokensStore(ctx)
+	swapFees := sdk.ZeroDec()
+
+	// Join/Exit pool support
+	if len(denomPath) == 1 {
```
[the remark this answers is not in the document]
Comment (lostak) [reply x/txfees/keeper/feetokens.go:60]: When I started adding support for join/exit with swap msgs, `TokenDenomsOnPath` had to be satisfied and return an array. 
However, this array would only hold the 1 denom for the token in/out and the loop relies on there being 2 denoms per pool id.

**Comment by lostak** [comment]
@p0mvn Thanks for your review, I've made the changes you've suggested but I still need to add the tests. 
Is there any documentation that you'd suggest to help me better understand how to use the testing suite?

**Comment by lostak** [comment]
Github's tests were failing because they were not up to date with the main branch.
I moved the changes to a new branch off the up to date main and the tests are passing now.
The new PR is #1823

**Review by Naser531900 — APPROVED** [review]

**Pull request closed without merging**